iceoryx2-services/discovery/src/service_discovery/service.rs

```diff
@@ -0,0 +1,371 @@
+            history_size: 10,
+            max_borrrowed_samples: 10,
+            send_notifications: true,
+            max_listeners: 10,
+        }
+    }
+}
+
+/// The service discovery service.
+///
+/// This service is responsible for tracking and publishing information about
+/// services in the system. It can detect when services are added or removed,
+/// and notify interested parties about these changes.
+///
+/// # Type Parameters
+///
+/// * `S` - The service type that this discovery service operates on.
+pub struct Service<S: ServiceType> {
+    discovery_config: Config,
+    iceoryx_config: IceoryxConfig,
+    _node: Node<S>,
+    publisher: Option<Publisher<S, [u8], ()>>,
+    notifier: Option<Notifier<S>>,
+    tracker: Tracker<S>,
+}
+
+impl<S: ServiceType> Service<S> {
+    /// Creates the service discovery service.
+    ///
+    /// This function initializes a new service discovery service with the provided
+    /// configuration. The service will track services of type `S` in the system.
+    ///
+    /// # Parameters
+    ///
+    /// * `discovery_config` - Configuration for the discovery service.
+    /// * `iceoryx_config` - Configuration for the underlying iceoryx system.
+    ///
+    /// # Returns
+    ///
+    /// A result containing either the created service or an error if creation failed.
+    ///
+    /// # Errors
+    ///
+    pub fn create(
+        discovery_config: &Config,
+        iceoryx_config: &IceoryxConfig,
+    ) -> Result<Self, CreationError> {
+        let node = NodeBuilder::new()
+            .config(iceoryx_config)
+            .create::<S>()
+            .map_err(|_| CreationError::NodeCreationFailure)?;
+
+        let mut publisher = None;
+        if discovery_config.publish_events {
+            let publish_subscribe = node
+                .service_builder(service_name())
+                .publish_subscribe::<Payload>()
+                .subscriber_max_buffer_size(discovery_config.max_buffer_size)
+                .subscriber_max_borrowed_samples(discovery_config.max_borrrowed_samples)
+                .history_size(discovery_config.history_size)
+                .max_subscribers(discovery_config.max_subscribers)
+                .max_publishers(1)
+                .open_or_create()
+                .map_err(|_| CreationError::ServiceCreationFailure)?;
+
+            publisher = Some(
+                publish_subscribe
+                    .publisher_builder()
+                    .initial_max_slice_len(1024)
+                    .allocation_strategy(AllocationStrategy::PowerOfTwo)
+                    .create()
+                    .map_err(|e| {match e {
+                        iceoryx2::port::publisher::PublisherCreateError::ExceedsMaxSupportedPublishers => CreationError::PublisherAlreadyExists,
+                        iceoryx2::port::publisher::PublisherCreateError::UnableToCreateDataSegment => CreationError::PublisherCreationError,
+                    }})?
```

Comment:
Food for thought. If there are `match` arms in `map_err`, it might be better to implement the `From` trait to be able to just use `?`. This moves the boilerplate out of sight.

```suggestion
                    .create()?
```

and at lind 68 this

```rs
use iceoryx2::port::publisher::PublisherCreateError;

impl From<PublisherCreateError> for CreationError {
    fn from(error: PublisherCreateError) -> Self {
        match error {
            PublisherCreateError::ExceedsMaxSupportedPublishers => {
                CreationError::PublisherAlreadyExists
            }
            PublisherCreateError::UnableToCreateDataSegment => {
                CreationError::PublisherCreationError
            }
        }
    }
}
```

@elfenpiff has some macros for this but I'm not quite sure how to use them. There is also #13 for further discussion.

Reply:
Tedious but done...